### PR TITLE
[DTensor] Add util API to compute_local_shape_and_global_offset for checkpointing purpose 

### PR DESCRIPTION
**Pull request by wz337**
The compute_local_shape_and_global_offset API does the following:
1) Calculate both local_shape and global_offset in one API to replace two API calls (compute_local_size and compute_local_shape).
2) Generate the correct global_offset for checkpointing purposes. We are currently using compute_local_offset for downstream checkpoint components, which could lead to incorrect results. For checkpointing, we need global_offset instead of local_offset. In some cases, global_offset does not equal to local_offset, when a dimension is sharded multipe times on different mesh dimension (e.g. placements = [Shard(0), Shard(0)]). 

Follow-up PRs:
1) Replace related downstream components to use compute_local_shape_and_global_offset instead of compute_local_size and compute_local_offset.
2) Audit existing code base to see if we can remove compute_local_size and compute_local_offset, since they are currently being used.

cc. @wanchaol

### CODE REVIEW

File: torch/distributed/_tensor/_utils.py [hunk 1]
```diff
@@ -77,6 +79,73 @@ def compute_local_offset(
         return tuple(local_offsets)
 
 
+def compute_local_shape_and_global_offset(
```
Comment (wanchaol) [inline torch/distributed/_tensor/_utils.py:82]: I think the original two APIs are correlated: 1. compute_local_tensor_shape 2. compute_local_tensor_offset, the second is not related to any "local_offset" per-say, it should actually compute the global offset of its "local shard", maybe the names are misleading. Shall we consolidate this API into the above two in this PR?

File: torch/distributed/_tensor/_utils.py [hunk 2]
```diff
@@ -77,6 +79,73 @@ def compute_local_offset(
...
+
+        for idx, placement in enumerate(placements):
+            mesh_dim_size = mesh.size(idx)
+            if isinstance(placement, Shard):
+                shard_dim = placement.dim
+                local_offset = [0] * len(global_shape)
+                assert shard_dim < len(
+                    local_shape
+                ), f"Sharding dim {shard_dim} greater than tensor ndim {len(local_shape)}"
+                shard_size, shard_offset = placement._local_shard_size_on_dim(
```
Comment (wanchaol) [inline torch/distributed/_tensor/_utils.py:127]: is this `shard_offset` be the offset on the mesh dimension in particular, and we just need to calculate the global offset base on this shard offset?

**Review by wanchaol — APPROVED** [review]
LGTM! As discussed offline, let's delete the compute_local_offset API in a follow up PR/diff

**Comment by wz337** [comment]
> LGTM! As discussed offline, let's delete the compute_local_offset API in a follow up PR/diff

Thanks for the review and discussion. Will follow up to delete compute_local_offset and migrate existing calls to use the new APIs in follow-up PRs.

**Pull request closed without merging**